tools/clang/lib/CodeGen/CGHLSLMS.cpp

```diff
@@ -5773,106 +5756,128 @@ Value *CGMSHLSLRuntime::EmitHLSLInitListExpr(CodeGenFunction &CGF, InitListExpr
-                            QualType Type, CodeGenTypes &Types,
-                            bool bDefaultRowMajor) {
+static void FlatConstToList(CodeGenTypes &Types, bool bDefaultRowMajor,
+    Constant *C, QualType QualTy,
+    SmallVectorImpl<Constant *> &EltVals, SmallVectorImpl<QualType> &EltQualTys) {
   llvm::Type *Ty = C->getType();
-  if (llvm::VectorType *VT = dyn_cast<llvm::VectorType>(Ty)) {
-    // Type is only for matrix. Keep use Type to next level.
-    for (unsigned i = 0; i < VT->getNumElements(); i++) {
-      FlatConstToList(C->getAggregateElement(i), EltValList, Type, Types,
-                      bDefaultRowMajor);
+  if (llvm::VectorType *VecTy = dyn_cast<llvm::VectorType>(Ty)) {
+    DXASSERT(hlsl::IsHLSLVecType(QualTy), "QualType/Type mismatch!");
+    for (unsigned i = 0; i < VecTy->getNumElements(); i++) {
+      FlatConstToList(Types, bDefaultRowMajor,
+        C->getAggregateElement(i), hlsl::GetHLSLVecElementType(QualTy),
+        EltVals, EltQualTys);
     }
   } else if (HLMatrixLower::IsMatrixType(Ty)) {
-    bool isRowMajor = IsRowMajorMatrix(Type, bDefaultRowMajor);
+    DXASSERT(hlsl::IsHLSLMatType(QualTy), "QualType/Type mismatch!");
     // matrix type is struct { vector<Ty, row> [col] };
     // Strip the struct level here.
-    Constant *matVal = C->getAggregateElement((unsigned)0);
-    const RecordType *RT = Type->getAs<RecordType>();
-    RecordDecl *RD = RT->getDecl();
-    QualType EltTy = RD->field_begin()->getType();
-    // When scan, init list scalars is row major.
-    if (isRowMajor) {
-      // Don't change the major for row major value.
-      FlatConstToList(matVal, EltValList, EltTy, Types, bDefaultRowMajor);
-    } else {
-      // Save to tmp list.
-      SmallVector<Constant *, 4> matEltList;
-      FlatConstToList(matVal, matEltList, EltTy, Types, bDefaultRowMajor);
-      unsigned row, col;
-      HLMatrixLower::GetMatrixInfo(Ty, col, row);
-      // Change col major value to row major.
-      for (unsigned r = 0; r < row; r++)
-        for (unsigned c = 0; c < col; c++) {
-          unsigned colMajorIdx = c * row + r;
-          EltValList.emplace_back(matEltList[colMajorIdx]);
+    Constant *RowArrayVal = C->getAggregateElement((unsigned)0);
+    QualType MatEltQualTy = hlsl::GetHLSLMatElementType(QualTy);
+
+    unsigned RowCount, ColCount;
+    hlsl::GetHLSLMatRowColCount(QualTy, RowCount, ColCount);
+
+    // Get all the elements from the array of row vectors.
+    // Matrices are never in memory representation so convert as needed.
+    SmallVector<Constant *, 16> MatElts;
+    for (unsigned r = 0; r < RowCount; ++r) {
+      Constant *RowVec = RowArrayVal->getAggregateElement(r);
+      for (unsigned c = 0; c < ColCount; ++c) {
+        Constant *MatElt = RowVec->getAggregateElement(c);
+        if (MatEltQualTy->isBooleanType()) {
+          DXASSERT(MatElt->getType()->isIntegerTy(1),
+            "Matrix elements should be in their register representation.");
```

Comment:
True, but I'd rather not do so because it would lead to a really ugly assert condition like `!MatEltQualTy->isBooleanType() || RowArrayVal->getAggregateElement(0)->getAggregateElement(0)->getType()->isIntegerType(1)`.